File: Pawn.py
from __future__ import annotations

from Piece import Piece, BoardGrid
from Cell import Cell
from Queen import Queen
from constants import PieceValue


class Pawn(Piece):
    def __init__(self, row: int, col: int, team) -> None:
        self.row = row
        self.col = col
        self.targets: list[Cell] = []
        self.team = team
        self.starting_row = row
        self.starting_col = col
        if self.row == 6:
            self.direction = -1
        else:
            self.direction = 1
        self.en_passantable: bool = False
        self.en_passant_loc: Cell | None = None
        self.critical: bool = False
        self.critical_targets: list[Cell] = []

    def calc_targets(self, board: BoardGrid) -> None:
        self.en_passant_loc = None
        self.targets = []
        if self.row != 0 and self.row != 7:
            if self.col != 0 and isinstance(board[self.row + self.direction][self.col - 1], Piece):
                if board[self.row + self.direction][self.col - 1].team != self.team:
                    self.targets.append(Cell(self.row + self.direction, self.col - 1))

            if self.col != 7 and isinstance(board[self.row + self.direction][self.col + 1], Piece):
                if board[self.row + self.direction][self.col + 1].team != self.team:
                    self.targets.append(Cell(self.row + self.direction, self.col + 1))

            if board[self.row + self.direction][self.col] is None:
                self.targets.append(Cell(self.row + self.direction, self.col))

            if self.row == self.starting_row:
                if (board[self.row + 2 * self.direction][self.col] is None
                        and board[self.row + self.direction][self.col] is None):
                    self.targets.append(Cell(self.row + 2 * self.direction, self.col))

            if (self.col != 0
                    and isinstance(board[self.row][self.col - 1], Pawn)
                    and board[self.row][self.col - 1].team != self.team):
                if board[self.row][self.col - 1].en_passantable:
                    self.targets.append(Cell(self.row + self.direction, self.col - 1))
                    self.en_passant_loc = Cell(self.row + self.direction, self.col - 1)

            if (self.col != 7
                    and isinstance(board[self.row][self.col + 1], Pawn)
                    and board[self.row][self.col + 1].team != self.team):
                if board[self.row][self.col + 1].en_passantable:
                    self.targets.append(Cell(self.row + self.direction, self.col + 1))
                    self.en_passant_loc = Cell(self.row + self.direction, self.col + 1)

        if self.critical:
            super().filter_to_pin_ray()

    def get_value(self, board: BoardGrid) -> int:
        self.calc_targets(board)
        total = PieceValue.PAWN
        total = total + len(self.targets)
        for cell in self.targets:
            total += 1
            if (cell.row == 3 or cell.row == 4) and (cell.col == 3 or cell.col == 4):
                total += 1
        return total

    def filter_to_king_escape(self, king) -> None:
        new_targets = []
        for target in self.targets:
            for saving_target in king.king_escape_cells:
                if target.row == saving_target.row and target.col == saving_target.col:
                    new_targets.append(target)
        if self.en_passant_loc is not None:
            new_targets.append(Cell(self.en_passant_loc.row, self.en_passant_loc.col))
        self.targets = new_targets

    def move(self, new_row: int, new_col: int, board: BoardGrid) -> Cell | None:
        old_row = self.row
        old_col = self.col
        self.row = new_row
        self.col = new_col
        if (self.starting_row + 6) % 12 == self.row:
            board[self.row][self.col] = Queen(self.row, self.col, self.team)
        if (old_row == self.starting_row
                and old_col == self.starting_col
                and (new_row - old_row) == 2 * self.direction):
            self.en_passantable = True
        if (self.en_passant_loc is not None
                and new_row == self.en_passant_loc.row
                and new_col == self.en_passant_loc.col):
            return Cell(self.en_passant_loc.row - self.direction, self.en_passant_loc.col)
        else:
            return None

    def print_piece(self) -> None:
        print("Pawn at", self.row, ",", self.col)

File: Chess101iOS/Sources/Chess101Engine/Pawn.swift
public final class Pawn: Piece {
    public var row: Int
    public var col: Int
    public var team: Team
    public var targets: [Cell] = []
    public var touched: Bool = false
    public var critical: Bool = false
    public var criticalTargets: [Cell] = []

    /// Direction of travel: +1 (started row 1) or -1 (started row 6).
    /// Note: direction is set by starting row, not team — BUG-01 preserved.
    public var direction: Int
    public var startingRow: Int
    public var startingCol: Int
    public var enPassantable: Bool = false
    public var enPassantLoc: Cell? = nil

    public var typeName: String { "Pawn" }

    public init(row: Int, col: Int, team: Team) {
        self.row = row
        self.col = col
        self.team = team
        self.startingRow = row
        self.startingCol = col
        self.direction = (row == 6) ? -1 : 1
    }

    @discardableResult
    public func calcTargets(board: Board) -> Bool? {
        targets = []
        enPassantLoc = nil  // Reset stale location from previous turns — matches Python
        let r = row, c = col, d = direction

        // One square forward
        if r + d >= 0 && r + d < 8 && board.grid[r + d][c] == nil {
            targets.append(Cell(r + d, c))
            // Two squares forward only from the starting row — matches Python (uses row, not touched)
            if r == startingRow && r + 2 * d >= 0 && r + 2 * d < 8 && board.grid[r + 2 * d][c] == nil {
                targets.append(Cell(r + 2 * d, c))
            }
        }

        // Diagonal captures
        for dc in [-1, 1] {
            let nr = r + d, nc = c + dc
            guard nr >= 0 && nr < 8 && nc >= 0 && nc < 8 else { continue }
            if let target = board.grid[nr][nc], target.team != team {
                targets.append(Cell(nr, nc))
            }
            // En passant
            if nc >= 0 && nc < 8, let neighbor = board.grid[r][nc] as? Pawn,
               neighbor.team != team, neighbor.enPassantable {
                targets.append(Cell(nr, nc))
                enPassantLoc = Cell(nr, nc)
            }
        }

        if critical { criticalMan() }
        return nil
    }

    public func move(toRow: Int, toCol: Int, board: Board) -> MoveResult {
        let wasEnPassant = enPassantLoc == Cell(toRow, toCol) && board.grid[toRow][toCol] == nil
        let fromRow = row  // save before mutation
        row = toRow; col = toCol
        // Note: do NOT set touched = true — Python Pawn.move() preserves touched=False (matches hash)
        // Set enPassantable only when double-stepping FROM the starting row — matches Python
        enPassantable = (fromRow == startingRow && abs(toRow - fromRow) == 2)
        if wasEnPassant {
            return .enPassantCapture(capturedAt: Cell(fromRow, toCol))
        }
        return .normal
    }

    /// Pawn-specific skyFall: also allows en passant when the captured pawn is the checker.
    public func skyFall(king: King) {
        let originalTargets = targets
        var newTargets = targets.filter { king.godSaveTheKing.contains($0) }

        // En passant special case: the captured pawn is at (enPassantLoc.row - direction, col),
        // not at the destination square — so the standard filter misses it.
        if let loc = enPassantLoc {
            let capturedRow = loc.row - direction
            let capturesChecker = king.godSaveTheKing.contains(Cell(capturedRow, loc.col))
            if capturesChecker && originalTargets.contains(loc) && !newTargets.contains(loc) {
                newTargets.append(loc)
            }
        }
        targets = newTargets
    }

    public func getValue(board: Board) -> Int {
        var val = 5
        // Center bonus
        let cr = abs(row - 3) <= 1 ? 1 : 0
        let cc = abs(col - 3) <= 1 ? 1 : 0
        val += cr + cc
        return val
    }
}
